README.md
# Artron - multiprocess execution helper

![GitHub](https://img.shields.io/github/license/ahmet2mir/python-artron.svg)
[![image](https://img.shields.io/pypi/pyversions/artron.svg)](https://python.org/pypi/artron)
[![Build Status](https://travis-ci.org/ahmet2mir/python-artron.svg?branch=master)](https://travis-ci.org/ahmet2mir/python-artron)

`Artron` is a simple `python` module to use `multiprocessing` with dependency graph and queue management allowing easy tool creation.

[Read the full documentation](https://artron.readthedocs.io/en/latest/) to see everything you can do.

## Installation

pip

    pip install artron

## Simple example

> see examples/basic.py to get the full file

First create an class with all actions needed inside (you could also use a module). `Artron` will use `getattr` on the object or module to run the task.

```python
import time

class Builder(object):
    
    def builder_func_1(self, msg, retry):
        time.sleep(6)
        return "builder_func_1 ==> " + msg

    def builder_func_2(self, msg, retry):
        time.sleep(3)
        return "builder_func_2 ==> " + msg

    def builder_func_3(self, msg, retry):
        time.sleep(1)
        raise Exception("ERROR builder_func_3")
        return "builder_func_3 ==> " + msg

    def builder_func_4(self, msg, retry):
        time.sleep(4)
        return "builder_func_4 ==> " + msg
```

Now use with Artron

```python
import sys
from pprint import pprint
from artron.task import Task
from artron.manager import Manager

def process(run=True):
    # builder to pass to Artron
    builder = Builder()

    # declare manager
    manager = Manager(builder, max_retry=2)

    # declare tasks
    task1 = Task('task-id-1', {'msg': 'task-1-msg'}, 'builder_func_1')
    task2 = Task('task-id-2', {'msg': 'task-2-msg'}, 'builder_func_2')
    task3 = Task('task-id-3', {'msg': 'task-3-msg'}, 'builder_func_3')
    task4 = Task('task-id-4', {'msg': 'task-4-msg'}, 'builder_func_4')
    task5 = Task('task-id-5', {'msg': 'task-5-msg'}, 'builder_func_2')
    task6 = Task('task-id-6', {'msg': 'task-6-msg'}, 'builder_func_4')
    
    # generate dependencies
    task1.add_require(task2.tid)
    task1.add_require(task3.tid)
    task1.add_require(task4.tid)
    task2.add_require(task4.tid)
    task6.add_require(task2.tid)
    task5.add_require(task1.tid)
    
    # add task
    manager.add(task1)
    manager.add(task2)
    manager.add(task3)
    manager.add(task4)
    manager.add(task5)
    manager.add(task6)

    # if we should run
    if run:
        # start
        results = manager.start()
        pprint(results)
        sys.exit(results['exit_code'])
    # otherwise print tasks ids
    else:
        for task in manager.tasks.keys():
            print("-", task)

if __name__ == '__main__':
    process()
```

That's it!

**Result**

```shell
artron: [ERROR] Traceback (most recent call last):
  File "/home/ahmet/workspaces/devs/python-artron/artron/task.py", line 89, in run
    **self.inputs
  File "examples/basic.py", line 19, in builder_func_3
    raise Exception("ERROR builder_func_3")
Exception: ERROR builder_func_3

artron: [ERROR] ERROR builder_func_3
artron: [ERROR] Traceback (most recent call last):
  File "/home/ahmet/workspaces/devs/python-artron/artron/task.py", line 89, in run
    **self.inputs
  File "examples/basic.py", line 19, in builder_func_3
    raise Exception("ERROR builder_func_3")
Exception: ERROR builder_func_3

artron: [ERROR] ERROR builder_func_3
{'date_end': '2018-08-29T20:02:54.640Z',
 'date_start': '2018-08-29T20:02:39.606Z',
 'elapsed': '00:00:15',
 'exit_code': 1,
 'results': {'aborted': 0,
             'deps': 2,
             'failures': 1,
             'nrun': 0,
             'ready': 0,
             'success': 3},
 'tasks': [{'date_created': '2018-08-29T20:02:39.605Z',
            'date_end': None,
            'date_start': None,
            'func': 'builder_func_1',
            'inputs': {'msg': 'task-1-msg'},
            'require': ['task-id-2', 'task-id-4'],
            'results': None,
            'state': -2,
            'tid': 'task-id-1',
            'time_duration': 0.0,
            'time_duration_str': '00:00:00'},
           {'date_created': '2018-08-29T20:02:39.605Z',
            'date_end': '2018-08-29T20:02:41.617Z',
            'date_start': '2018-08-29T20:02:40.615Z',
            'func': 'builder_func_3',
            'inputs': {'msg': 'task-3-msg'},
            'require': [],
            'results': 'ERROR builder_func_3',
            'state': -1,
            'tid': 'task-id-3',
            'time_duration': 1.0019969940185547,
            'time_duration_str': '00:00:01'},
           {'date_created': '2018-08-29T20:02:39.605Z',
            'date_end': '2018-08-29T20:02:48.625Z',
            'date_start': '2018-08-29T20:02:45.621Z',
            'func': 'builder_func_2',
            'inputs': {'msg': 'task-2-msg'},
            'require': [],
            'results': 'builder_func_2 ==> task-2-msg',
            'state': 3,
            'tid': 'task-id-2',
            'time_duration': 3.0032620429992676,
            'time_duration_str': '00:00:03'},
           {'date_created': '2018-08-29T20:02:39.605Z',
            'date_end': None,
            'date_start': None,
            'func': 'builder_func_2',
            'inputs': {'msg': 'task-5-msg'},
            'require': [],
            'results': None,
            'state': -2,
            'tid': 'task-id-5',
            'time_duration': 0.0,
            'time_duration_str': '00:00:00'},
           {'date_created': '2018-08-29T20:02:39.605Z',
            'date_end': '2018-08-29T20:02:43.616Z',
            'date_start': '2018-08-29T20:02:39.613Z',
            'func': 'builder_func_4',
            'inputs': {'msg': 'task-4-msg'},
            'require': [],
            'results': 'builder_func_4 ==> task-4-msg',
            'state': 3,
            'tid': 'task-id-4',
            'time_duration': 4.003551006317139,
            'time_duration_str': '00:00:04'},
           {'date_created': '2018-08-29T20:02:39.605Z',
            'date_end': '2018-08-29T20:02:54.636Z',
            'date_start': '2018-08-29T20:02:50.631Z',
            'func': 'builder_func_4',
            'inputs': {'msg': 'task-6-msg'},
            'require': [],
            'results': 'builder_func_4 ==> task-6-msg',
            'state': 3,
            'tid': 'task-id-6',
            'time_duration': 4.00443696975708,
            'time_duration_str': '00:00:04'}]}
```

## Details

### Tasks

`Task` is an object with `function` to run on `builder`, `inputs` to send and `state` also with time informations.

**States values:**

* -2: dependency error, means that parent task failed.
* -1: task failed to run
* 0: initial state (when creating them)
* 1: task is ready and in queue, there is no more dependencies, waiting to be process by workers
* 2: task is running
* 3: task run successfully

### Worker

`Worker` is just a `process`. 
By default, `multiprocessing.cpu_count()` is used as number of workers.

## Integrations

You could use [click](http://click.pocoo.org/5/) to create simple cli.

```shell
pip install click
```

```python
# -*- coding: utf-8 -*-
import basic # import examples/basic.py

import click

@click.group()
@click.version_option()
def cli():
    """Artron CLI"""

@cli.command()
def start():
    """Start all tasks."""
    basic.process()

@cli.command()
def list():
    """List tasks."""
    basic.process(False)

if __name__ == '__main__':
    cli()
```

List tasks

```shell
$ python examples/basic_cli.py list
- task-id-1
- task-id-3
- task-id-2
- task-id-5
- task-id-4
- task-id-6
```

# Licence

Copyright 2018 - Ahmet Demir

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.



docs/changelog.rst
=========
Changelog
=========

v0.0.4 - 25/10/2018
===================
- Add python versions to setup.py

v0.0.3 - 25/10/2018
===================
- Fix CI and doc

v0.0.2 - 25/10/2018
===================
- Add tests
- Add CI
- Add ProgressBar, both of tqdm and custom
- Fix ProxyDict access with RLock

v0.0.1 - 29/08/2018
===================
- First release


artron/manager.py
# -*- coding: utf-8 -*-
"""
atron.manager
~~~~~~~~~~~~~
"""
# standard
import os
import time
import multiprocessing

import logging
import logging.config
import logging.handlers

# local
from artron import utils
from artron.task import Task
from artron.graph import Graph
from artron.worker import Worker
from artron._py6 import range_type, TimeoutError


logging.config.dictConfig({
    'version': 1,
    'disable_existing_loggers': False,
    'formatters': {
        'simple': {
            '()': 'logging.Formatter',
            'format': 'artron: [%(levelname)s] %(message)s'
        }
    },
    'handlers': {
        'console':{
            'level': os.environ.get('ARTRON_LEVEL', 'ERROR').upper(),
            'class': 'logging.StreamHandler',
            'formatter': 'simple'
        }
    },
    'loggers': {
        'artron': {
            'level': os.environ.get('ARTRON_LEVEL', 'ERROR').upper(),
            'handlers': ['console']
        }
    }
})
LOGGER = logging.getLogger(__name__)


class Manager(object): # pylint: disable=too-many-instance-attributes
    """Manager class is the central piece of Artron.
    It takes care of creating, start and queueing jobs.

    Attributes:
        builder (obj): Builder object with the `func` to run.
        lock (multiprocessing.Lock): Lock on ressource access.
        max_retry (int): Number of retry when task fail.
        progress (obj): Progress bar.
        queue (multiprocessing.Manager.Queue): queue can be shared between
            subprocesses.
        sleep (int): sleep value in seconds when iterating edges.
        tasks (multiprocessing.managers.DictProxy): shared dict object and
            return a proxy for it.
        timeout (int): timestamp when timeout will be triggered.
        workers (list):  list of `artron.worker.Worker`

    Args:
        builder (obj): Builder object with the `func` to run.
        nb_workers (int): number of worker
        workers (list):  list of `artron.worker.Worker`
        queue (multiprocessing.Manager.Queue): queue can be shared between
            subprocesses.
        run_timeout (int): number of seconds for timeout. Will be added to
            attribute timeout with current timestamp.
        sleep (int): sleep value itasks (multiprocessing.managers.DictProxy):
            shared dict object and return a proxy for it.tasks=None
        max_retry (int): Number of retry when task fail.
        progress (obj): Progress bar.


    Examples:
        >>> manager = Manager(builder, max_retry=2)
    """
    # pylint: disable=too-many-arguments
    def __init__(self, builder, nb_workers=None, workers=None, queue=None, \
                 run_timeout=3600, sleep=1, tasks=None, max_retry=3, \
                 progress=None):
        self.builder = builder
        self.nb_workers = nb_workers
        self.queue = queue
        self.tasks = tasks
        self.workers = workers
        self.timeout = time.time() + run_timeout
        self.sleep = sleep
        self.max_retry = max_retry
        self.progress = progress

        if self.nb_workers is None:
            self.nb_workers = multiprocessing.cpu_count()

        mng = multiprocessing.Manager()

        if self.queue is None:
            self.queue = mng.Queue()

        if self.tasks is None:
            self.tasks = mng.dict()

        # set lock on tasks access
        self.lock = mng.RLock()

        if self.workers is None:
            self.workers = [
                Worker(
                    self.builder,
                    self.queue,
                    "worker-%d" % wid,
                    self.tasks,
                    max_retry,
                    self.lock,
                )
                for wid in range_type(self.nb_workers)
            ]

    def add(self, task):
        """Add task to manager

        Args:
            task (artron.task.Task): task object o add
            retry (bool): retry count
        """
        with self.lock:
            self.tasks[task.tid] = task

    # pylint: disable=too-many-branches,too-many-statements
    def start(self):
        """Start manager

        Returns:
            dict: run result in form.
                >>> {
                    'date_end': '2018-08-29T20:02:54.640Z',
                    'date_start': '2018-08-29T20:02:39.606Z',
                    'elapsed': '00:00:15',
                    'exit_code': 1,
                    'results': {
                        'aborted': 0,
                        'deps': 2,
                        'failures': 1,
                        'nrun': 0,
                        'ready': 0,
                        'success': 3
                    },
                    'tasks': [
                        {
                            'date_created': '2018-08-29T20:02:39.605Z',
                            'date_end': None,
                            'date_start': None,
                            'func': 'builder_func_1',
                            'inputs': {'msg': 'task-1-msg'},
                            'require': ['task-id-2', 'task-id-4'],
                            'results': None,
                            'state': -2,
                            'tid': 'task-id-1',
                            'time_duration': 0.0,
                            'time_duration_str': '00:00:00'
                        },{
                            'date_created': '2018-08-29T20:02:39.605Z',
                            'date_end': '2018-08-29T20:02:41.617Z',
                            'date_start': '2018-08-29T20:02:40.615Z',
                            'func': 'builder_func_3',
                            'inputs': {'msg': 'task-3-msg'},
                            'require': [],
                            'results': 'ERROR builder_func_3',
                            'state': -1,
                            'tid': 'task-id-3',
                            'time_duration': 1.0019969940185547,
                            'time_duration_str': '00:00:01'
                        }
                    ]
                }
        """
        time_start = time.time()

        out = {
            'elapsed': 0.0,
            'date_end': None,
            'date_start': utils.strdate(),
            'results': {
                'success': 0,
                'failures': 0,
                'deps': 0,
                'nrun': 0,
                'aborted': 0,
                'ready': 0
            },
            'exit_code': 1,
            'tasks': []
        }

        try:
            # start all workers
            LOGGER.debug("init %d workers", self.nb_workers)
            for worker in self.workers:
                worker.start()

            LOGGER.debug("send resources to queues")

            # generate exec graph
            graph = Graph(self.tasks)
            edges = list(graph.edges())
            LOGGER.debug("Edges %s", edges)

            # while we have non empty graph and don't reach timeout
            while edges and time.time() < self.timeout:
                # isolated vertice is task without deps
                for task_id in graph.isolated_vertices():

                    LOGGER.debug("isolated vertex task(%s) with state(%d)", \
                        task_id, self.tasks[task_id].state)

                    if self.tasks[task_id].is_runnable():
                        LOGGER.debug("send task(%s)", task_id)

                        # update task status because put in queue != is running
                        # so to avoid multiple queue send, mark it as running
                        with self.lock:
                            task_new = self.tasks[task_id]
                            task_new.state = Task.STATE_READY
                            self.tasks[task_id] = task_new

                        self.queue.put((task_id,))

                graph = Graph(self.tasks)
                edges = list(graph.edges())

                # LOGGER.debug("Task %s", self.tasks)
                LOGGER.debug("Edges %s", edges)

                if self.progress:
                    count = max(0, len([1 for task in self.tasks.values() \
                        if task.is_finished()]) - self.progress.n)
                    if count > 0:
                        self.progress.update(count)

                time.sleep(self.sleep)

            if self.progress:
                count = max(0, len([1 for task in self.tasks.values() \
                    if task.is_finished()]) - self.progress.n)
                if count > 0:
                    self.progress.update(count)

            if time.time() > self.timeout:
                raise TimeoutError('timeout error')

            LOGGER.debug("add end-of-queue markers")
            for _ in self.workers:
                # True add the end to mark the end of queue
                self.queue.put((None,))

            LOGGER.debug("blocks until all items in the queue have been "\
                  "gotten and processed.")
            self.queue.join()

            if self.progress:
                self.progress.close()

        # catch all errors
        # pylint: disable=broad-except
        except Exception as err:
            LOGGER.error("%s. Exiting...", err)

        # stop workers
        finally:
            LOGGER.debug("stop all workers")
            for worker in self.workers:
                # send sigterm
                worker.stop()
                # wait end
                worker.join()
                LOGGER.debug("stop workers %s", worker.name)

        out['date_end'] = utils.strdate()

        # final message
        for task in self.tasks.values():
            out['tasks'].append(task.__dict__)
            if task.state == Task.STATE_SUCCESS:
                out['results']['success'] += 1

            elif task.state == Task.STATE_ERROR:
                out['results']['failures'] += 1

            elif task.state == Task.STATE_DEPENDENCY:
                out['results']['deps'] += 1

            elif task.state == Task.STATE_INIT:
                out['results']['nrun'] += 1

            elif task.state == Task.STATE_RUNNING:
                out['results']['aborted'] += 1

            elif task.state == Task.STATE_READY:
                out['results']['ready'] += 1

        if len(out['tasks']) == out['results']['success']:
            out['exit_code'] = 0

        out['elapsed'] = utils.strgmtime(time.gmtime(time.time() - time_start))

        return out
